### PR TITLE
Add InputObject parameter sets for appropriate *-PSResource cmdlets, to provide pipeline support

### LINKED ISSUE #482: Uninstall-PSResource using a -Version parameter removes all versions of the resource!
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PowerShellGet)
- [X] Search the existing issues.

### Steps to reproduce

I installed v3.0.11-beta in order to test the **fix** for uninstalling modules from a OneDrive location, [Issue# 300](https://github.com/PowerShell/PowerShellGet/issues/300). The Uninstall-PSResource cmdlet removed ALL versions of a module when only specifying a single version.

### Expected behavior

```console
C:\Users\JDoe>Get-InstalledPSResource -Name Az.WvdOperations

Name             Version   Prerelease Description
----             -------   ---------- -----------
Az.WvdOperations 2.21.9.8             AVD Operations Module
Az.WvdOperations 2.21.9.5             AVD Operations Module
Az.WvdOperations 2.21.8.23            AVD Operations Module
Az.WvdOperations 2.21.8.20            AVD Operations Module


C:\Users\JDoe>Get-InstalledPSResource -Name Az.WvdOperations -Version 2.21.8.20

Name             Version   Prerelease Description
----             -------   ---------- -----------
Az.WvdOperations 2.21.8.20            AVD Operations Module


Get-InstalledPSResource -Name Az.WvdOperations -Version 2.21.8.20 | Uninstall-PSResource

Only v2.21.8.20 would be removed
```

### Actual behavior

```console
C:\Users\JDoe>Get-InstalledPSResource -Name Az.WvdOperations

Name             Version   Prerelease Description
----             -------   ---------- -----------
Az.WvdOperations 2.21.9.8             AVD Operations Module
Az.WvdOperations 2.21.9.5             AVD Operations Module
Az.WvdOperations 2.21.8.23            AVD Operations Module
Az.WvdOperations 2.21.8.20            AVD Operations Module


C:\Users\JDoe>Get-InstalledPSResource -Name Az.WvdOperations -Version 2.21.8.20

Name             Version   Prerelease Description
----             -------   ---------- -----------
Az.WvdOperations 2.21.8.20            AVD Operations Module


Get-InstalledPSResource -Name Az.WvdOperations -Version 2.21.8.20 | Uninstall-PSResource

All versions removed!
```

### Error details

```console
No Errors
```

### Environment data

```powershell
ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Binary     3.0.11     PowerShellGet                       {Find-PSResource, Get-InstalledPSResource, Get-PSResourceRepository, Install-PSResource...}

Key   : PSVersion
Value : 5.1.19041.1151
Name  : PSVersion


Key   : PSEdition
Value : Desktop
Name  : PSEdition


Key   : PSCompatibleVersions
Value : {1.0, 2.0, 3.0, 4.0...}
Name  : PSCompatibleVersions


Key   : BuildVersion
Value : 10.0.19041.1151
Name  : BuildVersion


Key   : CLRVersion
Value : 4.0.30319.42000
Name  : CLRVersion


Key   : WSManStackVersion
Value : 3.0
Name  : WSManStackVersion


Key   : PSRemotingProtocolVersion
Value : 2.3
Name  : PSRemotingProtocolVersion


Key   : SerializationVersion
Value : 1.1.0.1
Name  : SerializationVersion
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by anamnavi**
# PR Summary

Add `InputObjectParameterSet` to Install-PSResource, Uninstall-PSResource, Save-PSResource.
Remove `ValueFromPipeline`/`ValueFromPipelineByPropertyName` for all but Name property for these cmdlets as well.

## PR Context

Add pipeline support, resolves #482 , #348 (partially).

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/InstallPSResource.cs [hunk 1]
```diff
@@ -54,44 +53,58 @@ class InstallPSResource : PSCmdlet
...
         /// For modules that require a license, AcceptLicense automatically accepts the license agreement during installation.
         /// </summary>
         [Parameter(ParameterSetName = NameParameterSet)]
+        [Parameter(ParameterSetName = InputObjectParameterSet)]
         public SwitchParameter AcceptLicense { get; set; }
 
+        /// <summary>
+        /// Used for pipeline input.
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ValueFromPipelineByPropertyName = true, ParameterSetName = InputObjectParameterSet)]
```
Comment (alerickson) [inline src/code/InstallPSResource.cs]: Is the ValueFromPipelineByPropertyName needed here?

**Review by alerickson — APPROVED** [review]

File: src/code/InstallPSResource.cs (hunk 1 above)
Comment (anamnavi) [reply src/code/InstallPSResource.cs]: good question, I checked and don't need that attribute. I've removed `ValueFromPipelineByPropertyName` from Install, Uninstall and Save

File: src/code/InstallPSResource.cs [hunk 2]
```diff
@@ -54,44 +53,58 @@ class InstallPSResource : PSCmdlet
...
         [Parameter(ParameterSetName = NameParameterSet)]
+        [Parameter(ParameterSetName = InputObjectParameterSet)]
         public SwitchParameter AcceptLicense { get; set; }
 
+        /// <summary>
+        /// Used for pipeline input.
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ParameterSetName = InputObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSResourceInfo[] InputObject { get; set; }
```
Comment (PaulHigin) [inline src/code/InstallPSResource.cs]:
```suggestion
        public PSResourceInfo InputObject { get; set; }

```
It is not necessary to take the InputObject parameter as an array type.

File: src/code/InstallPSResource.cs [hunk 3]
```diff
@@ -107,87 +120,60 @@ protected override void BeginProcessing()
...
-            }
-
             var installHelper = new InstallHelper(updatePkg: false, savePkg: false, cmdletPassedIn: this);
 
             switch (ParameterSetName)
             {
                 case NameParameterSet:
-                    var namesToInstall = Utils.ProcessNameWildcards(Name, out string[] errorMsgs, out bool nameContainsWildcard);
-                    if (nameContainsWildcard)
+                    // If no Version specified, install latest version for the package.
```
Comment (PaulHigin) [inline src/code/InstallPSResource.cs:132]: Please remove extra white space.

File: src/code/SavePSResource.cs [hunk 4]
```diff
@@ -109,15 +110,16 @@ public string Path
...
         public SwitchParameter TrustRepository { get; set; }
         
         /// <summary>
         /// Used for pipeline input.
         /// </summary>
-        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ValueFromPipelineByPropertyName = true, ParameterSetName = InputObjectSet)]
+        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ParameterSetName = InputObjectParameterSet)]
         [ValidateNotNullOrEmpty]
-        public object[] InputObject { set; get; }
+        public PSResourceInfo[] InputObject { get; set; }
```
Comment (PaulHigin) [inline src/code/SavePSResource.cs]: Array type not needed.

File: src/code/UninstallPSResource.cs [hunk 5]
```diff
@@ -39,19 +39,20 @@ public sealed class UninstallPSResource : PSCmdlet
         /// <summary>
         /// Used for pipeline input.
         /// </summary>
-        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ValueFromPipelineByPropertyName = true, ParameterSetName = InputObjectSet)]
+        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true, ParameterSetName = InputObjectParameterSet)]
```
Comment (PaulHigin) [inline src/code/UninstallPSResource.cs:42]: Array type not needed.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Looks good.  Just some minor suggestions.

**Review by PaulHigin — APPROVED** [review]

**Pull request merged**